lib/tapioca/compilers/dsl/active_support_concern.rb

```diff
@@ -0,0 +1,115 @@
+# typed: strict
+# frozen_string_literal: true
+
+require "tapioca/compilers/sorbet"
+
+begin
+  require "active_support"
+rescue LoadError
+  return
+end
+
+return unless Tapioca::Compilers::Sorbet.supports?(:mixes_in_class_methods_multiple_args)
+
+module Tapioca
+  module Compilers
+    module Dsl
+      # `Tapioca::Compilers::Dsl::ActiveSupportConcern` generates RBI files for classes that both `extend`
+      # `ActiveSupport::Concern` and `include` another class that extends `ActiveSupport::Concern`
+      #
+      # For example for the following hierarchy:
+      #
+      # ~~~rb
+      # # concern.rb
+      # module Foo
+      #  extend ActiveSupport::Concern
+      #  module ClassMethods; end
+      # end
+      #
+      # module Bar
+      #  extend ActiveSupport::Concern
+      #  module ClassMethods; end
+      #  include Foo
+      # end
+      #
+      # class Baz
+      #  include Bar
+      # end
+      # ~~~
+      #
+      # this generator will produce the RBI file `concern.rbi` with the following content:
+      #
+      # ~~~rbi
+      # # typed: true
+      # module Bar
+      #   mixes_in_class_methods(::Foo::ClassMethods)
+      # end
+      # ~~~
+      class ActiveSupportConcern < Base
+        extend T::Sig
+
+        sig { override.params(root: Parlour::RbiGenerator::Namespace, constant: Module).void }
+        def decorate(root, constant)
+          dependencies = linearized_dependencies_of(constant)
+
+          mixed_in_class_methods = dependencies
+            .uniq # Deduplicate
+            .map do |concern| # Map to class methods module name, if exists
+              "::#{name_of(concern)}::ClassMethods" if concern.const_defined?(:ClassMethods)
+            end
+            .compact # Remove non-existent records
+
+          return if mixed_in_class_methods.empty?
+
+          root.path(constant) do |mod|
+            mixin_module_list = mixed_in_class_methods.join(", ")
+
+            mod.create_arbitrary(code: "mixes_in_class_methods(#{mixin_module_list})")
+          end
+        end
+
+        sig { override.returns(T::Enumerable[Module]) }
+        def gather_constants
+          modules = T.cast(ObjectSpace.each_object(Module), T::Enumerable[Module])
+
+          # Find all Modules that are:
+          # - named (i.e. not anonymous)
+          # - not singleton classes
+          # - extend ActiveSupport::Concern, and
+          # - have dependencies (i.e. include another concern)
+          modules.select do |mod|
+            name_of(mod) &&
+            !mod.singleton_class? &&
+            mod.singleton_class < ActiveSupport::Concern &&
+            !dependencies_of(mod).empty?
```

Comment:
We certainly can, but I prefer not to since they are not exactly equivalent. `any?` could return `false` for an array that is not empty:
```ruby
![nil, false].empty? => true
# but
[nil, false].any? # => false
```

That optimization makes much more sense in the case of AR queries, since the query `! .. empty?` would generate would be different to the query `any?` would generate, and thus, you would end up loading less records. In this case, the array from the `dependencies_of(mod)` would have already been allocated anyway, so the difference does not matter and `!...empty?` is the safer version.

Reply:
Yes, that is very much intentional since that's the semantics of `any?`. That method is a test to see if there are any truthy elements in an enumerable. That is super obvious with the block form of `any?` which only returns `true` if any of the block calls return `true`. Same story with `all?`. People have co-opted `any?` to mean something different than what it actually means, thus the confusion.

Ruby core is having discussions around introducing a method that would exactly mean `!...empty?` but nothing concrete so far.